File: go/pkg/node/node.go
// Package node provides a high-level embeddable network node for zgrnet.
//
// Node wraps the low-level UDP transport, Noise Protocol encryption, and KCP
// stream multiplexing into a simple API suitable for apps, embedded devices,
// and Go services. Unlike Host, Node does not require TUN or root privileges.
//
// Basic usage:
//
//	n, err := node.New(node.Config{PrivateKey: keyPair, ListenPort: 0})
//	if err != nil { ... }
//	defer n.Stop()
//
//	n.AddPeer(node.PeerConfig{PublicKey: remotePK, Endpoint: "1.2.3.4:51820"})
//
//	stream, err := n.Dial(remotePK, 8080)
//	// or
//	stream, err := n.AcceptStream()
package node

import (
	"errors"
	"fmt"
	"net"
	"sync"
	"sync/atomic"
	"time"

	"github.com/vibing/zgrnet/pkg/kcp"
	znet "github.com/vibing/zgrnet/pkg/net"
	"github.com/vibing/zgrnet/pkg/noise"
	"github.com/vibing/zgrnet/pkg/relay"
)

// State represents the lifecycle state of a Node.
type State int32

const (
	StateStopped State = iota
	StateRunning
	StateSuspended
)

func (s State) String() string {
	switch s {
	case StateStopped:
		return "stopped"
	case StateRunning:
		return "running"
	case StateSuspended:
		return "suspended"
	default:
		return "unknown"
	}
}

// Errors returned by Node operations.
var (
	ErrNotRunning     = errors.New("node: not running")
	ErrAlreadyRunning = errors.New("node: already running")
	ErrPeerNotFound   = errors.New("node: peer not found")
	ErrNotConnected   = errors.New("node: peer not connected")
	ErrStopped        = errors.New("node: stopped")
)

// Config holds the configuration for creating a Node.
type Config struct {
	// PrivateKey is the Noise Protocol keypair. Required.
	PrivateKey *noise.KeyPair

	// ListenPort is the UDP port to listen on. 0 for OS-assigned.
	ListenPort int

	// AllowUnknown allows accepting connections from peers not added via AddPeer.
	AllowUnknown bool
}

// PeerConfig holds the configuration for adding a peer.
type PeerConfig struct {
	// PublicKey is the peer's Curve25519 public key. Required.
	PublicKey noise.PublicKey

	// Endpoint is the peer's UDP address in "host:port" format.
	// Empty means responder-only (wait for the peer to connect to us).
	Endpoint string
}

// Node is an embeddable zgrnet network node.
//
// It provides Dial (active connect), AcceptStream (passive accept),
// and raw UDP send/recv — all over Noise-encrypted KCP transport.
// No TUN device, no root privileges required.
type Node struct {
	config Config
	udp    *znet.UDP

	// Global accept channel aggregates streams from all peers.
	acceptCh chan *Stream

	// Tracks per-peer accept goroutines.
	peerMu   sync.Mutex
	peerDone map[noise.PublicKey]chan struct{} // signal to stop per-peer accept loop

	// Lifecycle
	state atomic.Int32
	done  chan struct{}
	wg    sync.WaitGroup
}

// New creates a new Node but does not start it.
// Call Start() to begin listening and accepting connections.
func New(cfg Config) (*Node, error) {
	if cfg.PrivateKey == nil {
		return nil, errors.New("node: PrivateKey is required")
	}

	rt := relay.NewRouteTable()

	bindAddr := fmt.Sprintf("127.0.0.1:%d", cfg.ListenPort)
	udp, err := znet.NewUDP(
		cfg.PrivateKey,
		znet.WithBindAddr(bindAddr),
		znet.WithAllowUnknown(cfg.AllowUnknown),
		znet.WithRouteTable(rt),
	)
	if err != nil {
		return nil, fmt.Errorf("node: create UDP: %w", err)
	}

	n := &Node{
		config:   cfg,
		udp:      udp,
		acceptCh: make(chan *Stream, 64),
		peerDone: make(map[noise.PublicKey]chan struct{}),
		done:     make(chan struct{}),
	}
	n.state.Store(int32(StateRunning))

	// Start the background receive loop.
	// This is required for handshakes and transport messages to be processed.
	n.wg.Add(1)
	go func() {
		defer n.wg.Done()
		n.recvLoop()
	}()

	return n, nil
}

// Stop shuts down the Node and releases all resources.
func (n *Node) Stop() {
	if !n.state.CompareAndSwap(int32(StateRunning), int32(StateStopped)) &&
		!n.state.CompareAndSwap(int32(StateSuspended), int32(StateStopped)) {
		return // already stopped
	}

	close(n.done)

	// Stop all per-peer accept loops.
	n.peerMu.Lock()
	for _, ch := range n.peerDone {
		close(ch)
	}
	n.peerMu.Unlock()

	n.udp.Close()
	n.wg.Wait()
}

// State returns the current lifecycle state.
func (n *Node) State() State {
	return State(n.state.Load())
}

// PublicKey returns this node's public key.
func (n *Node) PublicKey() noise.PublicKey {
	return n.config.PrivateKey.Public
}

// LocalAddr returns the local UDP address the node is listening on.
func (n *Node) LocalAddr() net.Addr {
	return n.udp.HostInfo().Addr
}

// AddPeer registers a peer. If endpoint is provided, initiates a handshake.
func (n *Node) AddPeer(cfg PeerConfig) error {
	if n.State() != StateRunning {
		return ErrNotRunning
	}

	if cfg.Endpoint != "" {
		ep, err := net.ResolveUDPAddr("udp", cfg.Endpoint)
		if err != nil {
			return fmt.Errorf("node: resolve endpoint %q: %w", cfg.Endpoint, err)
		}
		n.udp.SetPeerEndpoint(cfg.PublicKey, ep)
	} else {
		// Register peer without endpoint (responder-only).
		n.udp.SetPeerEndpoint(cfg.PublicKey, &net.UDPAddr{})
	}

	// Start accept forwarder goroutine for this peer.
	n.startAcceptLoop(cfg.PublicKey)

	return nil
}

// RemovePeer removes a peer and closes all associated streams.
func (n *Node) RemovePeer(pk noise.PublicKey) {
	// Stop the accept loop for this peer.
	n.peerMu.Lock()
	if ch, ok := n.peerDone[pk]; ok {
		close(ch)
		delete(n.peerDone, pk)
	}
	n.peerMu.Unlock()

	n.udp.RemovePeer(pk)
}

// Peers returns information about all registered peers.
func (n *Node) Peers() []znet.PeerInfo {
	var result []znet.PeerInfo
	for p := range n.udp.Peers() {
		result = append(result, *p.Info)
	}
	return result
}

// Connect initiates a handshake with a peer. The peer must have been added
// via AddPeer with an endpoint. Blocks until the handshake completes or fails.
func (n *Node) Connect(pk noise.PublicKey) error {
	if n.State() != StateRunning {
		return ErrNotRunning
	}
	return n.udp.Connect(pk)
}

// Dial connects to a peer and opens a KCP stream.
//
// pk identifies the remote peer. port is the target port carried as metadata
// in the stream's SYN frame (proto=TCP_PROXY, address=127.0.0.1:port).
//
// If the peer is not yet connected, Dial automatically initiates a handshake
// and waits for it to complete before opening the stream.
func (n *Node) Dial(pk noise.PublicKey, port uint16) (*Stream, error) {
	if n.State() != StateRunning {
		return nil, ErrNotRunning
	}

	// Ensure peer is connected.
	info := n.udp.PeerInfo(pk)
	if info == nil {
		return nil, ErrPeerNotFound
	}
	if info.State != znet.PeerStateEstablished {
		if err := n.udp.Connect(pk); err != nil {
			return nil, fmt.Errorf("node: connect to %s: %w", pk.ShortString(), err)
		}
	}

	// Build target address metadata.
	addr := &noise.Address{
		Type: noise.AddressTypeIPv4,
		Host: "127.0.0.1",
		Port: port,
	}
	metadata := addr.Encode()

	raw, err := n.udp.OpenStream(pk, noise.ProtocolKCP, metadata)
	if err != nil {
		return nil, fmt.Errorf("node: open stream to %s:%d: %w", pk.ShortString(), port, err)
	}

	return &Stream{Stream: raw, remotePK: pk}, nil
}

// OpenStream opens a raw KCP stream to a peer with custom proto and metadata.
// Use Dial for the common case of TCP_PROXY streams.
func (n *Node) OpenStream(pk noise.PublicKey, proto byte, metadata []byte) (*Stream, error) {
	if n.State() != StateRunning {
		return nil, ErrNotRunning
	}

	info := n.udp.PeerInfo(pk)
	if info == nil {
		return nil, ErrPeerNotFound
	}
	if info.State != znet.PeerStateEstablished {
		if err := n.udp.Connect(pk); err != nil {
			return nil, fmt.Errorf("node: connect: %w", err)
		}
	}

	raw, err := n.udp.OpenStream(pk, proto, metadata)
	if err != nil {
		return nil, err
	}
	return &Stream{Stream: raw, remotePK: pk}, nil
}

// DialRelay connects to a remote peer through a relay and opens a KCP stream.
//
// relayPK is the public key of the relay node that both this node and the
// remote peer are connected to. The relay node forwards handshake and data
// packets between the two peers (via RELAY_0/RELAY_2).
//
// The relay route is registered in the RouteTable, so all subsequent traffic
// to dst automatically goes through the relay (including KCP stream data).
//
// Both this node and the relay must have established sessions (call AddPeer +
// Connect for the relay first).
func (n *Node) DialRelay(dst noise.PublicKey, relayPK noise.PublicKey, port uint16) (*Stream, error) {
	if n.State() != StateRunning {
		return nil, ErrNotRunning
	}

	// Register relay route: dst → relay
	n.udp.RouteTable().AddRoute(dst, relayPK)

	// Register the peer without an endpoint (relay-only).
	// AddPeer is idempotent if the peer is already registered.
	n.AddPeer(PeerConfig{PublicKey: dst})

	// Dial goes through initiateHandshake (relay-aware) then OpenStream.
	return n.Dial(dst, port)
}

// RouteTable returns the node's relay route table.
func (n *Node) RouteTable() *relay.RouteTable {
	return n.udp.RouteTable()
}

// AcceptStream waits for an incoming KCP stream from any peer.
// Returns the stream with RemotePubkey() identifying the sender.
func (n *Node) AcceptStream() (*Stream, error) {
	select {
	case s := <-n.acceptCh:
		return s, nil
	case <-n.done:
		return nil, ErrStopped
	}
}

// WriteTo sends raw data to a peer with the given protocol byte.
// This is a connectionless send (no stream, no reliability).
func (n *Node) WriteTo(data []byte, protocol byte, pk noise.PublicKey) error {
	if n.State() != StateRunning {
		return ErrNotRunning
	}
	_, err := n.udp.Write(pk, protocol, data)
	return err
}

// ReadFrom reads the next raw packet from any peer.
// Returns the number of bytes read, the protocol byte, and the sender's public key.
func (n *Node) ReadFrom(buf []byte) (int, byte, noise.PublicKey, error) {
	pk, proto, n2, err := n.udp.ReadPacket(buf)
	return n2, proto, pk, err
}

// UDP returns the underlying UDP transport. Advanced use only.
func (n *Node) UDP() *znet.UDP {
	return n.udp
}

// recvLoop runs in the background to drive the UDP receive pipeline.
// Without this, handshakes and transport messages won't be processed.
func (n *Node) recvLoop() {
	buf := make([]byte, 65535)
	for {
		select {
		case <-n.done:
			return
		default:
		}
		_, _, err := n.udp.ReadFrom(buf)
		if err != nil {
			if errors.Is(err, znet.ErrClosed) {
				return
			}
			continue
		}
	}
}

// startAcceptLoop starts a goroutine that forwards accepted streams
// from one peer into the global acceptCh.
func (n *Node) startAcceptLoop(pk noise.PublicKey) {
	n.peerMu.Lock()
	if _, exists := n.peerDone[pk]; exists {
		n.peerMu.Unlock()
		return // already running
	}
	stopCh := make(chan struct{})
	n.peerDone[pk] = stopCh
	n.peerMu.Unlock()

	n.wg.Add(1)
	go func() {
		defer n.wg.Done()
		n.acceptLoopForPeer(pk, stopCh)
	}()
}

// acceptLoopForPeer reads accepted streams from a specific peer and
// pushes them to the global accept channel.
//
// It first waits for the peer to reach PeerStateEstablished (polling at 50ms),
// then enters the blocking AcceptStream loop.
func (n *Node) acceptLoopForPeer(pk noise.PublicKey, stop <-chan struct{}) {
	// Phase 1: Wait for the peer to establish a session.
	// AcceptStream returns ErrNoSession immediately if the peer isn't
	// connected yet, so we poll PeerInfo to avoid a busy loop.
	for {
		info := n.udp.PeerInfo(pk)
		if info != nil && info.State == znet.PeerStateEstablished {
			break
		}
		select {
		case <-stop:
			return
		case <-n.done:
			return
		case <-time.After(50 * time.Millisecond):
			// Retry.
		}
	}

	// Phase 2: Accept streams. AcceptStream blocks until a stream arrives
	// or the UDP instance is closed, so this is not a busy loop.
	for {
		raw, err := n.udp.AcceptStream(pk)
		if err != nil {
			if errors.Is(err, znet.ErrClosed) || errors.Is(err, znet.ErrPeerNotFound) {
				return
			}
			// Session may have been reset (rekey, reconnect). Go back to
			// waiting for re-establishment.
			select {
			case <-stop:
				return
			case <-n.done:
				return
			case <-time.After(50 * time.Millisecond):
				continue
			}
		}

		s := &Stream{Stream: raw, remotePK: pk}
		select {
		case n.acceptCh <- s:
		case <-stop:
			raw.Close()
			return
		case <-n.done:
			raw.Close()
			return
		}
	}
}

// Stream wraps a KCP stream with the remote peer's public key.
type Stream struct {
	*kcp.Stream
	remotePK noise.PublicKey
}

// RemotePubkey returns the public key of the peer on the other end.
func (s *Stream) RemotePubkey() noise.PublicKey {
	return s.remotePK
}

File: rust/src/node.rs
//! Embeddable zgrnet network node without TUN.
//!
//! `Node` wraps the low-level UDP transport, Noise Protocol encryption, and KCP
//! stream multiplexing into a simple API suitable for apps and services.
//! Unlike `Host`, Node does not require TUN or root privileges.
//!
//! # Example
//!
//! ```rust,ignore
//! use zgrnet::node::{Node, NodeConfig, PeerConfig};
//! use zgrnet::noise::KeyPair;
//!
//! let key = KeyPair::generate();
//! let node = Node::new(NodeConfig { key, listen_port: 0, allow_unknown: true })?;
//! node.add_peer(PeerConfig { public_key: remote_pk, endpoint: Some("1.2.3.4:51820".into()) })?;
//! let stream = node.dial(&remote_pk, 8080)?;
//! ```

use crate::kcp::SyncStream;
use crate::net::{PeerInfo, PeerState, UdpError, UdpOptions, UDP};
use crate::noise::{Key, KeyPair};
use crate::relay::RouteTable;

use crossbeam_channel::{bounded, Receiver, Sender};
use std::collections::HashMap;
use std::fmt;
use std::io;
use std::net::SocketAddr;
use std::sync::atomic::{AtomicU8, Ordering};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

/// Lifecycle state of a Node.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum State {
    Stopped = 0,
    Running = 1,
    Suspended = 2,
}

impl fmt::Display for State {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            State::Stopped => write!(f, "stopped"),
            State::Running => write!(f, "running"),
            State::Suspended => write!(f, "suspended"),
        }
    }
}

/// Errors from Node operations.
#[derive(Debug)]
pub enum NodeError {
    NotRunning,
    AlreadyRunning,
    PeerNotFound,
    NotConnected,
    Stopped,
    Udp(UdpError),
    Io(std::io::Error),
    Other(String),
}

impl fmt::Display for NodeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            NodeError::NotRunning => write!(f, "node: not running"),
            NodeError::AlreadyRunning => write!(f, "node: already running"),
            NodeError::PeerNotFound => write!(f, "node: peer not found"),
            NodeError::NotConnected => write!(f, "node: peer not connected"),
            NodeError::Stopped => write!(f, "node: stopped"),
            NodeError::Udp(e) => write!(f, "node: udp: {}", e),
            NodeError::Io(e) => write!(f, "node: io: {}", e),
            NodeError::Other(s) => write!(f, "node: {}", s),
        }
    }
}

impl std::error::Error for NodeError {}

impl From<UdpError> for NodeError {
    fn from(e: UdpError) -> Self {
        NodeError::Udp(e)
    }
}

impl From<std::io::Error> for NodeError {
    fn from(e: std::io::Error) -> Self {
        NodeError::Io(e)
    }
}

type Result<T> = std::result::Result<T, NodeError>;

/// Configuration for creating a Node.
pub struct NodeConfig {
    /// Noise Protocol keypair. Required.
    pub key: KeyPair,
    /// UDP port to listen on. 0 for OS-assigned.
    pub listen_port: u16,
    /// Allow connections from peers not added via add_peer.
    pub allow_unknown: bool,
}

/// Configuration for adding a peer.
pub struct PeerConfig {
    /// Peer's Curve25519 public key.
    pub public_key: Key,
    /// Peer's UDP address in "host:port" format. None for responder-only.
    pub endpoint: Option<String>,
}

/// A yamux stream with the remote peer's public key and service ID attached.
pub struct NodeStream {
    pub stream: SyncStream,
    pub remote_pk: Key,
    pub service: u64,
}

impl NodeStream {
    pub fn remote_pubkey(&self) -> Key {
        self.remote_pk
    }

    pub fn read(&self, buf: &mut [u8]) -> io::Result<usize> {
        let mut s = self.stream.clone();
        io::Read::read(&mut s, buf)
    }

    pub fn write(&self, data: &[u8]) -> io::Result<usize> {
        let mut s = self.stream.clone();
        io::Write::write(&mut s, data)
    }

    pub fn close(&self) {
        self.stream.close();
    }
}

/// An embeddable zgrnet network node.
///
/// Provides Dial (active connect), AcceptStream (passive accept),
/// and raw UDP send/recv — all over Noise-encrypted KCP transport.
/// No TUN device, no root privileges required.
pub struct Node {
    udp: Arc<UDP>,
    state: AtomicU8,

    // Global accept channel aggregates streams from all peers.
    accept_tx: Sender<NodeStream>,
    accept_rx: Receiver<NodeStream>,

    // Per-peer stop signals for accept forwarder threads.
    peer_stops: Mutex<HashMap<Key, Sender<()>>>,

    // Keep track of threads for clean shutdown.
    stop_tx: Sender<()>,
    stop_rx: Receiver<()>,
}

impl Node {
    /// Creates and starts a new Node.
    pub fn new(cfg: NodeConfig) -> Result<Arc<Self>> {
        let bind_addr = format!("127.0.0.1:{}", cfg.listen_port);
        let udp = UDP::new(
            cfg.key,
            UdpOptions::new().bind_addr(&bind_addr).allow_unknown(cfg.allow_unknown),
        )
        .map_err(NodeError::Udp)?;

        let rt = Arc::new(RouteTable::new());
        udp.set_route_table(rt);

        let udp = Arc::new(udp);
        let (accept_tx, accept_rx) = bounded(64);
        let (stop_tx, stop_rx) = bounded(1);

        let node = Arc::new(Self {
            udp,
            state: AtomicU8::new(State::Running as u8),
            accept_tx,
            accept_rx,
            peer_stops: Mutex::new(HashMap::new()),
            stop_tx,
            stop_rx,
        });

        // Start background receive loop.
        let n = Arc::clone(&node);
        thread::spawn(move || n.recv_loop());

        Ok(node)
    }

    /// Shuts down the Node and releases all resources.
    pub fn stop(&self) {
        let prev = self.state.swap(State::Stopped as u8, Ordering::SeqCst);
        if prev == State::Stopped as u8 {
            return; // already stopped
        }

        // Signal stop to all accept forwarder threads.
        let _ = self.stop_tx.try_send(());
        let mut stops = self.peer_stops.lock().unwrap();
        for (_, tx) in stops.drain() {
            let _ = tx.try_send(());
        }

        self.udp.close().ok();
    }

    /// Returns the current lifecycle state.
    pub fn state(&self) -> State {
        match self.state.load(Ordering::SeqCst) {
            1 => State::Running,
            2 => State::Suspended,
            _ => State::Stopped,
        }
    }

    /// Returns this node's public key.
    pub fn public_key(&self) -> Key {
        self.udp.host_info().public_key
    }

    /// Returns the local UDP address.
    pub fn local_addr(&self) -> SocketAddr {
        self.udp.host_info().addr
    }

    /// Registers a peer. If endpoint is provided, the peer can be connected to.
    pub fn add_peer(&self, cfg: PeerConfig) -> Result<()> {
        if self.state() != State::Running {
            return Err(NodeError::NotRunning);
        }

        if let Some(ref ep_str) = cfg.endpoint {
            let ep: SocketAddr = ep_str
                .parse()
                .map_err(|e| NodeError::Other(format!("resolve endpoint: {}", e)))?;
            self.udp.set_peer_endpoint(cfg.public_key, ep);
        } else {
            // Register peer without endpoint.
            self.udp
                .set_peer_endpoint(cfg.public_key, "0.0.0.0:0".parse().unwrap());
        }

        // Start accept forwarder thread for this peer.
        self.start_accept_loop(cfg.public_key);

        Ok(())
    }

    /// Removes a peer.
    pub fn remove_peer(&self, pk: &Key) {
        // Stop the accept forwarder.
        let mut stops = self.peer_stops.lock().unwrap();
        if let Some(tx) = stops.remove(pk) {
            let _ = tx.try_send(());
        }
        drop(stops);

        self.udp.remove_peer(pk);
    }

    /// Returns information about all registered peers.
    pub fn peers(&self) -> Vec<PeerInfo> {
        self.udp.peers().into_iter().map(|p| p.info).collect()
    }

    /// Initiates a handshake with a peer.
    pub fn connect(&self, pk: &Key) -> Result<()> {
        if self.state() != State::Running {
            return Err(NodeError::NotRunning);
        }
        self.udp.connect(pk).map_err(NodeError::from)
    }

    /// Connects to a peer and opens a yamux stream on the given service.
    pub fn dial(&self, pk: &Key, service: u64) -> Result<NodeStream> {
        if self.state() != State::Running {
            return Err(NodeError::NotRunning);
        }

        let info = self.udp.peer_info(pk).ok_or(NodeError::PeerNotFound)?;
        if info.state != PeerState::Established {
            self.udp.connect(pk).map_err(NodeError::from)?;
        }

        let stream = self.udp.open_stream(pk, service).map_err(NodeError::from)?;
        Ok(NodeStream { stream, remote_pk: *pk, service })
    }

    /// Connects to a remote peer through a relay and opens a yamux stream.
    pub fn dial_relay(&self, dst: &Key, relay_pk: &Key, service: u64) -> Result<NodeStream> {
        if self.state() != State::Running {
            return Err(NodeError::NotRunning);
        }

        if let Some(rt) = self.udp.route_table() {
            rt.add_route(dst.0, relay_pk.0);
        }

        self.add_peer(PeerConfig { public_key: *dst, endpoint: None })?;
        self.dial(dst, service)
    }

    /// Returns the node's relay route table.
    pub fn route_table(&self) -> Option<Arc<RouteTable>> {
        self.udp.route_table()
    }

    /// Waits for an incoming yamux stream from any peer.
    pub fn accept_stream(&self) -> Result<NodeStream> {
        self.accept_rx
            .recv()
            .map_err(|_| NodeError::Stopped)
    }

    /// Sends raw data to a peer with the given protocol byte (no stream).
    pub fn write_to(&self, data: &[u8], protocol: u8, pk: &Key) -> Result<()> {
        if self.state() != State::Running {
            return Err(NodeError::NotRunning);
        }
        self.udp
            .write_to_protocol(pk, protocol, data)
            .map_err(NodeError::from)
    }

    /// Reads the next raw packet from any peer.
    /// Returns (bytes_read, protocol, sender_pk).
    pub fn read_from(&self, buf: &mut [u8]) -> Result<(usize, u8, Key)> {
        let (pk, proto, n) = self.udp.read_packet(buf).map_err(NodeError::from)?;
        Ok((n, proto, pk))
    }

    /// Returns the underlying UDP transport. Advanced use only.
    pub fn udp(&self) -> &UDP {
        &self.udp
    }

    // ── Internal ──────────────────────────────────────────────────────

    /// Background receive loop — drives the UDP receive pipeline.
    fn recv_loop(&self) {
        let mut buf = vec![0u8; 65535];
        loop {
            if self.state() == State::Stopped {
                return;
            }
            match self.udp.read_from(&mut buf) {
                Ok(_) => {}
                Err(UdpError::Closed) => return,
                Err(_) => continue,
            }
        }
    }

    /// Starts a thread that forwards accepted streams from one peer
    /// into the global accept channel.
    fn start_accept_loop(&self, pk: Key) {
        let mut stops = self.peer_stops.lock().unwrap();
        if stops.contains_key(&pk) {
            return; // already running
        }
        let (stop_tx, stop_rx) = bounded(1);
        stops.insert(pk, stop_tx);
        drop(stops);

        let udp = Arc::clone(&self.udp);
        let accept_tx = self.accept_tx.clone();
        let global_stop = self.stop_rx.clone();

        thread::spawn(move || {
            // Phase 1: Wait for peer to become established.
            loop {
                if let Some(info) = udp.peer_info(&pk) {
                    if info.state == PeerState::Established {
                        break;
                    }
                }
                // Check stop signals.
                if stop_rx.try_recv().is_ok() || global_stop.try_recv().is_ok() {
                    return;
                }
                thread::sleep(Duration::from_millis(50));
            }

            // Phase 2: Accept streams.
            loop {
                if stop_rx.try_recv().is_ok() || global_stop.try_recv().is_ok() {
                    return;
                }

                match udp.accept_stream(&pk) {
                    Ok((stream, service)) => {
                        let ns = NodeStream { stream, remote_pk: pk, service };
                        if accept_tx.send(ns).is_err() {
                            return;
                        }
                    }
                    Err(UdpError::Closed) | Err(UdpError::PeerNotFound) => return,
                    Err(_) => {
                        // Transient error (session reset, etc.) — retry after delay.
                        thread::sleep(Duration::from_millis(50));
                    }
                }
            }
        });
    }
}

impl Drop for Node {
    fn drop(&mut self) {
        self.stop();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::noise::KeyPair;

    fn gen_key(seed: u8) -> KeyPair {
        let mut private = [0u8; 32];
        private[0] = seed;
        KeyPair::from_private(Key::from(private))
    }

    #[test]
    fn test_new_and_stop() {
        let kp = KeyPair::generate();
        let node = Node::new(NodeConfig {
            key: kp,
            listen_port: 0,
            allow_unknown: false,
        })
        .unwrap();
        assert_eq!(node.state(), State::Running);
        node.stop();
        assert_eq!(node.state(), State::Stopped);
    }

    #[test]
    fn test_double_stop() {
        let kp = KeyPair::generate();
        let node = Node::new(NodeConfig {
            key: kp,
            listen_port: 0,
            allow_unknown: false,
        })
        .unwrap();
        node.stop();
        node.stop(); // should not panic
    }

    #[test]
    fn test_add_peer_and_peers() {
        let kp1 = gen_key(1);
        let kp2 = gen_key(2);
        let node = Node::new(NodeConfig {
            key: kp1,
            listen_port: 0,
            allow_unknown: false,
        })
        .unwrap();

        node.add_peer(PeerConfig {
            public_key: kp2.public,
            endpoint: Some("127.0.0.1:19999".into()),
        })
        .unwrap();

        let peers = node.peers();
        assert_eq!(peers.len(), 1);
        assert_eq!(peers[0].public_key, kp2.public);

        node.stop();
    }

    #[test]
    fn test_remove_peer() {
        let kp1 = gen_key(1);
        let kp2 = gen_key(2);
        let node = Node::new(NodeConfig {
            key: kp1,
            listen_port: 0,
            allow_unknown: false,
        })
        .unwrap();

        node.add_peer(PeerConfig {
            public_key: kp2.public,
            endpoint: Some("127.0.0.1:19999".into()),
        })
        .unwrap();
        node.remove_peer(&kp2.public);

        assert_eq!(node.peers().len(), 0);
        node.stop();
    }

    #[test]
    fn test_operations_on_stopped_node() {
        let kp = KeyPair::generate();
        let node = Node::new(NodeConfig {
            key: kp,
            listen_port: 0,
            allow_unknown: false,
        })
        .unwrap();
        node.stop();

        assert!(node.add_peer(PeerConfig {
            public_key: Key::default(),
            endpoint: None,
        }).is_err());
        assert!(node.connect(&Key::default()).is_err());
        assert!(node.dial(&Key::default(), 1).is_err());
        assert!(node.write_to(&[], 0, &Key::default()).is_err());
    }
}
